File: go/pkg/edge/edgerecord/range_string_corpus_test.go
package edgerecord

import (
	"errors"
	"net/netip"
	"strings"
	"testing"

	edgev1 "github.com/carverauto/serviceradar/proto/edge/v1"
	"google.golang.org/protobuf/proto"
)

// Task 1.5-h: PLAN RANGE ADDRESS STRINGS -- the zone prohibition, and the guard bound that
// prohibition puts back out of reach.
//
// THREE PARTS, and only the first is a semantic rule:
//
//  1. ZONE-PRESENT REFUSAL, per field.
//  2. ACCEPTED CONTROLS -- at the frozen ceiling for the PREFLIGHT, and at the largest valid
//     syntax for the whole validator. Without these an always-refusing implementation
//     satisfies every refusal row.
//  3. ONE OVER-LIMIT, parser-not-entered control per field.
//
// THE CEILING IS PINNED AT THE SEAM, NOT THROUGH THE VALIDATOR. No canonical address reaches
// 64 bytes once zones are forbidden, so a whole-validator at-64 acceptance does not exist to
// construct -- and a 65-byte refusal driven through the validator survives the bound drifting
// anywhere from 43 to 65, since the parser refuses those lengths anyway. Only the preflight
// can distinguish them, so that is where 64 and 65 are asserted.

// THE SEAM IS TESTED DIRECTLY. `checkRangeStrings` parses nothing, so a row against it shows
// what the PREDICATES do and pins the frozen ceiling, without matching on DIAGNOSTIC WORDING
// that no requirement owns.
//
// IT DOES NOT PROVE ATTACHMENT. The checked value is an ordinary same-package struct, so a
// caller here can forge one and skip the preflight entirely. What catches that is the
// WHOLE-VALIDATOR zoned row below; in Elixir it is the traced stage test.
func TestRangeStringPreflightPinsTheFrozenLiteral(t *testing.T) {
	// LITERAL-PINNED. Building the vectors from the constant makes them move with it, so a
	// bound drifting to 43 or 128 would keep every row green. The frozen value is stated here
	// and the constant is checked against it.
	if MaxRangeStrBytes != 64 {
		t.Fatalf("MaxRangeStrBytes = %d, want the frozen 64", MaxRangeStrBytes)
	}

	for _, field := range []string{"cidr", "first", "last"} {
		t.Run(field, func(t *testing.T) {
			// AT the ceiling: accepted by the preflight. It is not a canonical address and the
			// parser will refuse it later -- which is the point: this row is about the PREFLIGHT,
			// and it is what stops a tightened `>` to `>=` from passing unnoticed.
			if _, err := checkRangeStrings(rangeWith(field, strings.Repeat("z", 64))); err != nil {
				t.Fatalf("exactly 64 bytes must pass the preflight: %v", err)
			}
			// ONE OVER: refused, before any parser is entered.
			if _, err := checkRangeStrings(rangeWith(field, strings.Repeat("z", 65))); !errors.Is(err, ErrPlanRange) {
				t.Fatalf("65 bytes = %v, want ErrPlanRange", err)
			}
		})
	}
}

func TestRangeStringPreflightRefusesZonesPerField(t *testing.T) {
	// Each field is an independently removable predicate: a prohibition applied to `cidr`
	// alone leaves a zoned span admitted. Tested at the seam, so the row does not depend on
	// what the whole validator would have done with the rest of the range.
	for _, field := range []string{"cidr", "first", "last"} {
		t.Run(field, func(t *testing.T) {
			if _, err := checkRangeStrings(rangeWith(field, "fe80::1%eth0")); !errors.Is(err, ErrPlanRange) {
				t.Fatalf("a zoned %s = %v, want ErrPlanRange", field, err)
			}
			// The SAME value without its zone passes the preflight, so the refusal is
			// attributable to the zone and not to anything else about the string.
			if _, err := checkRangeStrings(rangeWith(field, "fe80::1")); err != nil {
				t.Fatalf("the unzoned form must pass the preflight: %v", err)
			}
		})
	}
}

func TestZonedAddressIsAdmittedWithoutTheRule(t *testing.T) {
	// WHY THE RULE EXISTS, checked rather than argued. Without the preflight a zoned address
	// is canonical and parseable HERE, so this runtime would admit it -- and because a zone is
	// arbitrary-length text it reaches the ceiling exactly, which is what makes that bound
	// attainable at all.
	addr := "fe80::1%" + strings.Repeat("e", 64-len("fe80::1%"))
	if len(addr) != 64 {
		t.Fatalf("control failed: built %d bytes, want 64", len(addr))
	}

	a, err := netip.ParseAddr(addr)
	if err != nil {
		t.Fatalf("the address parser rejects the scoped form, so the rule guards nothing: %v", err)
	}
	if a.String() != addr {
		t.Fatal("the scoped form does not round-trip; the spelling rule would already catch it")
	}
}

func TestZonedRangeIsRefusedThroughTheWholeValidator(t *testing.T) {
	// THE SEAM ROWS ARE NOT ENOUGH ON THEIR OWN. `checkedRangeStrings` is an ordinary
	// same-package struct, so anything inside this package can FORGE one and reach the parser
	// without the preflight -- the handoff makes the order hard to get wrong, not impossible.
	// This row drives a zoned range through the real validator, resealed, so a bypass added
	// later fails here even though every seam row would still pass.
	//
	// EQUAL ENDPOINTS deliberately: a zoned span whose first and last are the same address is
	// the shape that would otherwise be ADMITTED, since it parses, round-trips canonically and
	// spans exactly one target.
	r := zoneCorpusRange(t)
	r.Cidr = ""
	r.FirstAddress, r.LastAddress = "fe80::1%eth0", "fe80::1%eth0"
	r.TargetCount = 1
	r.RangeSha256 = RangeDigest(r)

	if err := validateTargetRange(r, r.GetCheckSetSha256(), r.GetAvailabilityPolicyId()); !errors.Is(err, ErrPlanRange) {
		t.Fatalf("a zoned equal-endpoint span = %v, want ErrPlanRange", err)
	}

	// THE ACCEPTED CONTROL for the same shape. Without it an always-refusing validator
	// satisfies the row above, and the negative would prove only that something is wrong.
	ok := zoneCorpusRange(t)
	ok.Cidr = ""
	ok.FirstAddress, ok.LastAddress = "fe80::1", "fe80::1"
	ok.TargetCount = 1
	ok.RangeSha256 = RangeDigest(ok)

	if err := validateTargetRange(ok, ok.GetCheckSetSha256(), ok.GetAvailabilityPolicyId()); err != nil {
		t.Fatalf("the same span WITHOUT its zone must be accepted: %v", err)
	}
}

func TestPlanRangeAcceptsTheLargestValidSyntax(t *testing.T) {
	// The WHOLE-VALIDATOR acceptance controls. Without these an always-refusing validator
	// satisfies every refusal row above, and the preflight would look proven while the range
	// path admitted nothing.
	for _, tc := range []struct {
		name  string
		apply func(*edgev1.TargetRangeV1) uint64
	}{
		{"longest canonical CIDR", func(r *edgev1.TargetRangeV1) uint64 {
			p := netip.MustParsePrefix("ffff:ffff:ffff:ffff:ffff:ffff:ffff:8000/113")
			r.Cidr, r.FirstAddress, r.LastAddress = p.String(), "", ""
			return 1 << (p.Addr().BitLen() - p.Bits())
		}},
		{"longest canonical address span", func(r *edgev1.TargetRangeV1) uint64 {
			a := netip.MustParseAddr("ffff:ffff:ffff:ffff:ffff:ffff:ffff:ffff")
			r.Cidr = ""
			r.FirstAddress, r.LastAddress = a.String(), a.String()
			return 1
		}},
	} {
		t.Run(tc.name, func(t *testing.T) {
			r := zoneCorpusRange(t)
			r.TargetCount = tc.apply(r)
			r.RangeSha256 = RangeDigest(r)

			if err := validateTargetRange(r, r.GetCheckSetSha256(), r.GetAvailabilityPolicyId()); err != nil {
				t.Fatalf("the largest VALID %s must be accepted: %v", tc.name, err)
			}
		})
	}
}

// rangeWith puts one value in one address field and leaves the others empty, so each row
// exercises exactly the predicate it names.
func rangeWith(field, value string) *edgev1.TargetRangeV1 {
	r := &edgev1.TargetRangeV1{}

	switch field {
	case "cidr":
		r.Cidr = value
	case "first":
		r.FirstAddress = value
	case "last":
		r.LastAddress = value
	}

	return r
}

func zoneCorpusRange(t *testing.T) *edgev1.TargetRangeV1 {
	t.Helper()

	_, _, pages := planBoundAssignment(t)

	return proto.Clone(pages[0].GetRanges()[0]).(*edgev1.TargetRangeV1)
}

File: elixir/serviceradar_core/test/serviceradar/edge/range_string_corpus_test.exs
defmodule ServiceRadar.Edge.RangeStringCorpusTest do
  @moduledoc """
  Task 1.5-h: PLAN RANGE ADDRESS STRINGS -- the IPv6-zone prohibition, and the guard bound that
  prohibition puts back out of reach. Peer of Go's `range_string_corpus_test.go`.

  ## Three parts, and only the first is a semantic rule

    1. ZONE-PRESENT REFUSAL -- the only rows whose verdict changes if the prohibition is dropped.
    2. ACCEPTED CONTROLS at the largest valid syntax, per field. Without them an
       always-refusing validator satisfies part 3.
    3. ONE OVER-LIMIT, PARSER-NOT-ENTERED control per field.

  There is no WHOLE-VALIDATOR at-ceiling/one-over pair: once zones are forbidden no canonical
  address reaches `MaxRangeStrBytes`, so an input accepted at the ceiling by the whole
  validator does not exist to construct. The SEAM has such a pair, and that is where the
  frozen literal is pinned.

  ## WHERE EACH ROW PROVES SOMETHING, and where it only pins a verdict

  THE SEAM ROWS ARE MUTATION-KILLING: drifting the bound or tightening `>` to `>=` fails them,
  and so does dropping or narrowing the zone check. They are also the only place the frozen
  ceiling can be pinned here, since no canonical address reaches it.

  THE VALIDATOR-LEVEL ROWS ARE REGRESSION PINS ONLY, and that is a property of this runtime
  rather than a gap in the corpus. Everything the preflight refuses is ALSO refused downstream
  -- zones by the canonical-spelling check, over-length values by the parser -- so removing the
  preflight entirely, or replacing its call with a forged checked value, changes NO verdict.
  Measured across zoned spans, zoned CIDRs, over-length values in each field, and valid
  controls: every candidate is identical either way.

  THE ATTACHMENT IS PROVEN ELSEWHERE, by observing the STAGE rather than the verdict:
  `range_string_stage_test.exs` traces the preflight and `:inet.parse_strict_address/1` through
  the real validator, so a forged handoff fails there. No verdict row here can do that, and
  none pretends to.

  The reason is that this runtime already refuses a zoned address, by a different rule at a
  later stage. `:inet.parse_strict_address/1` ACCEPTS the zoned text and silently DISCARDS the
  zone, so `canonical_addr?/2` then compares `:inet.ntoa/1`'s output against the input and
  refuses the mismatch as a non-canonical SPELLING. No zoned value can round-trip, so the
  spelling rule catches every one -- and it reports `:plan_range`, the SAME atom the zone gate
  returns. A verdict assertion cannot separate them.

  The gate is still correct and still required: the frozen rule says a zone is refused BEFORE
  either parser, and the spelling rule runs after. THE STAGE IS OBSERVED AT THE VALIDATOR --
  `range_string_stage_test.exs` traces the preflight and `:inet.parse_strict_address/1` through
  the real `PlanValidate.validate/2` -- so no distinct refusal reason was ever needed. What no
  row in THIS file can do is separate the two by verdict.

  So these are REGRESSION rows: they pin the current verdict per field, and they would catch a
  change that made this runtime ADMIT a zoned address. They do not claim to prove the gate.
  Go's peer suite is where the gate is load-bearing and mutation-killed -- `netip` accepts a
  scoped address and round-trips it canonically, so without the gate Go ADMITS what this
  runtime refuses.
  """
  use ExUnit.Case, async: true

  alias ServiceRadar.Edge.HashGrammar
  alias ServiceRadar.Edge.PlanValidate

  @fixtures Path.expand("../../../../../proto/edge/v1/testdata", __DIR__)

  describe "the preflight seam" do
    # THE SEAM IS ADDRESSED DIRECTLY. `__check_range_strings__/3` parses nothing, so a row
    # against it constrains the ORDER -- an input refused there never reached a parser.
    # It is also the ONLY place the frozen ceiling can be pinned in this runtime: no canonical
    # address reaches 64 bytes, and a 65-byte refusal driven through `validate/2` survives the
    # bound drifting anywhere from 43 to 65, because the address parser refuses those lengths
    # regardless.

    test "the frozen literal 64 is pinned, and 65 is refused" do
      for field <- [:cidr, :first, :last] do
        assert {:ok, _} = with_field(field, String.duplicate("z", 64)),
               "exactly 64 bytes must pass the preflight for #{field}"

        assert {:error, :plan_range} = with_field(field, String.duplicate("z", 65)),
               "65 bytes must be refused for #{field}, before any parser"
      end
    end

    test "each field refuses a zone INDEPENDENTLY" do
      # A prohibition applied to `cidr` alone would leave a zoned span admitted.
      for field <- [:cidr, :first, :last] do
        assert {:error, :plan_range} = with_field(field, "fe80::1%eth0"),
               "a zoned #{field} must be refused"

        # The SAME value without its zone passes, so the refusal is attributable to the zone.
        assert {:ok, _} = with_field(field, "fe80::1")
      end
    end
  end

  describe "whole-validator acceptance" do
    test "the longest canonical CIDR is accepted" do
      # Without an accepted control an always-refusing implementation satisfies every refusal
      # row above, and the preflight would look proven while the range path admitted nothing.
      r = %{
        range()
        | cidr: "ffff:ffff:ffff:ffff:ffff:ffff:ffff:8000/113",
          first_address: "",
          last_address: ""
      }

      assert :ok = validate_range(%{r | target_count: 32_768})
    end

    test "the longest canonical address span is accepted" do
      addr = "ffff:ffff:ffff:ffff:ffff:ffff:ffff:ffff"
      r = %{range() | cidr: "", first_address: addr, last_address: addr}

      assert :ok = validate_range(%{r | target_count: 1})
    end

    test "a zoned address does not reach acceptance, and the reason stays :plan_range" do
      # A REGRESSION row for the VERDICT -- this runtime already refused zoned addresses
      # through its spelling check, so removing the gate does not change it. What it would
      # catch is a change that made this runtime ADMIT one.
      #
      # It also pins the REASON as the existing `:plan_range`. A fault-specific tag here
      # would be this subtask quietly minting a refusal class that 1.5-l owns.
      zoned = %{range() | cidr: "", first_address: "fe80::1%eth0", last_address: "fe80::1%eth0"}
      over = %{range() | cidr: String.duplicate("z", 65), first_address: "", last_address: ""}

      assert {:error, :plan_range} = validate_range(%{zoned | target_count: 1})
      assert {:error, :plan_range} = validate_range(%{over | target_count: 1})
    end
  end

  defp with_field(:cidr, v), do: PlanValidate.__check_range_strings__(v, "", "")
  defp with_field(:first, v), do: PlanValidate.__check_range_strings__("", v, "")
  defp with_field(:last, v), do: PlanValidate.__check_range_strings__("", "", v)

  # Drives the REAL page validator, resealing the range digest so a refusal is attributable to
  # the rule under test rather than to a stale digest.
  defp validate_range(r) do
    sealed = %{r | range_sha256: HashGrammar.range_digest(r)}
    page = %{page() | ranges: [sealed]}
    resealed = %{page | page_sha256: HashGrammar.plan_page_digest(page)}

    header = header_for(resealed)

    case PlanValidate.validate(header, [resealed]) do
      {:ok, _windows} -> :ok
      {:error, reason} -> {:error, reason}
    end
  end

  defp header_for(page) do
    # The MTR ordinal-range commitment is folded over the ranges, so changing a range without
    # recomputing it makes every row fail on the commitment instead of the rule under test.
    {:ok, commitment} = HashGrammar.plan_mtr_ordinal_range_commitment([page])

    h = %{
      header()
      | page_count: 1,
        plan_root_sha256: HashGrammar.plan_root([page]),
        total_target_count: hd(page.ranges).target_count,
        mtr_ordinal_range_commitment: commitment
    }

    %{h | execution_plan_sha256: HashGrammar.plan_header_digest(h)}
  end

  defp range, do: hd(page().ranges)

  defp page, do: "plan_page.bin" |> load() |> Serviceradar.Edge.V1.ScheduledPlanPageV1.decode()

  defp header,
    do: "plan_header.bin" |> load() |> Serviceradar.Edge.V1.ScheduledPlanHeaderV1.decode()

  defp load(name) do
    dir =
      [
        @fixtures,
        System.get_env("TEST_SRCDIR") &&
          Path.join([System.get_env("TEST_SRCDIR"), "_main", "proto/edge/v1/testdata"])
      ]
      |> Enum.reject(&is_nil/1)
      |> Enum.find(&File.dir?/1)

    if !dir, do: flunk("shared fixture directory not found")
    File.read!(Path.join(dir, name))
  end
end
